### PR TITLE
Rework snap files to handle refreshes

### PR DESCRIPTION
**Pull request by nrobinaubertin**
This extracts some logic out of the `install` hook to a `prepare` script that is also used by the new `post-refresh` hook.

### CODE REVIEW

File: service/bin/gunicorn-start [hunk 1]
```diff
@@ -9,4 +9,5 @@ export DJANGO_SECRET_KEY="$(tr -dc a-zA-Z0-9 < /dev/urandom | head -c 50)"
 export DJANGO_DEBUG="$(snapctl get django-debug)"
```
Comment (amandahla) [inline service/bin/gunicorn-start:9]: I'm not that familiar with Django/Gunicorn, but is it okay to set django_secret_key as the test key in the prepare step and have a different one here?

File: snap/hooks/install [hunk 2]
```diff
@@ -2,58 +2,33 @@
...
-# Copy nginx configuration
-cp -r --preserve=mode "$SNAP/nginx" "$SNAP_DATA/nginx"
-
-# Create some directories for nginx
-mkdir -p $SNAP_DATA/nginx/tmp/body
-mkdir -p $SNAP_DATA/nginx/tmp/nginx-proxy
-mkdir -p $SNAP_DATA/nginx/tmp/nginx-fastcgi
-mkdir -p $SNAP_DATA/nginx/tmp/nginx-uwsgi
-mkdir -p $SNAP_DATA/nginx/tmp/nginx-scgi
+cat "$SNAP_COMMON/bind/rndc.conf" | awk '/^# (Start|End|Use)/ {next} /^#/ {sub(/^# /, ""); print}' >> "$SNAP_COMMON/bind/named.conf"
```
Comment (amandahla) [inline snap/hooks/install:25]: awk wizard :mage:

File: service/bin/gunicorn-start (hunk 1 above)
Comment (nrobinaubertin) [reply service/bin/gunicorn-start:9]: It's just outside the diff, but there's a comment:
```
# The secret key is chosen randomly at the start of the gunicorn process
# It is not critical that it remains the same across restarts
# https://docs.djangoproject.com/en/5.1/ref/settings/#std:setting-SECRET_KEY
```

File: service/bin/prepare [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+# Create the static directory for django
+cp -r "$SNAP/api" "$SNAP_DATA/"
+chmod -R 755 $SNAP_DATA/api
+
+# Prepare the django app
+cd $SNAP_DATA/api
+export DJANGO_DATABASE_PATH="$SNAP_COMMON/api/db.sqlite3"
+export DJANGO_SECRET_KEY="testkey"
+python3 manage.py collectstatic --noinput
```
[the remark this answers is not in the document]
Comment (nrobinaubertin) [reply service/bin/prepare:55]: That's a mistake actually. good catch

[the remark this answers is not in the document]
Comment (nrobinaubertin) [reply service/bin/prepare:55]: (it should be python3 everywhere)

File: service/bin/prepare [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+# ----
+# NGINX
+# As nginx is only supplementary to the workload of the charm.
+# As such, its files are handled in SNAP_DATA
+
+# Copy nginx configuration
+cp -r --preserve=mode "$SNAP/nginx" "$SNAP_DATA/nginx"
+
+# Create some directories for nginx
+mkdir -p $SNAP_DATA/nginx/tmp/body
```
Comment (arturo-seijas) [inline service/bin/prepare:16]: Can you put quotes around these paths? It's considered good practice

File: service/bin/prepare [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+# Edit nginx configuration
+sed -i \
+  -e "s|<SNAP_DATA>|${SNAP_DATA}|g" \
+  -e "s|<SNAP_COMMON>|${SNAP_COMMON}|g" \
+  "$SNAP_DATA/nginx/nginx.conf"
+
+# Change ownership of some snap directories to allow snap_daemon to read/write
+# https://snapcraft.io/docs/system-usernames
+chown -R 584788:root $SNAP_DATA/nginx
```
Comment (arturo-seijas) [inline service/bin/prepare:30]: Same here

File: service/bin/prepare [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+# Copy the config files to SNAP_DATA
+cp -rf --preserve=mode "$SNAP/usr/share/dns" "$SNAP_DATA/usr/share"
+
+# ----
+# API (django and gunicorn)
+# The only thing that should be kept between refreshes is the database
+
+# Create the static directory for django
+cp -r "$SNAP/api" "$SNAP_DATA/"
+chmod -R 755 $SNAP_DATA/api
```
Comment (arturo-seijas) [inline service/bin/prepare:49]: Same here

File: service/bin/prepare [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+# Prepare the django app
+cd $SNAP_DATA/api
```
Comment (arturo-seijas) [inline service/bin/prepare:52]: I'd avoid cd-ing if possible

File: snap/hooks/install [hunk 8]
```diff
@@ -2,58 +2,33 @@
...
-mkdir -p "$SNAP_DATA/etc"
-mkdir -p "$SNAP_DATA/usr/share"
-mkdir -p "$SNAP_DATA/var/cache/bind"
+$SNAP/bin/prepare
 
-# Copy the config files to SNAP_DATA
-cp -rf --preserve=mode "$SNAP/etc/bind" "$SNAP_DATA/etc/bind"
-cp -rf --preserve=mode "$SNAP/usr/share/dns" "$SNAP_DATA/usr/share"
+# Create some directories
+mkdir -p $SNAP_COMMON/api
```
Comment (arturo-seijas) [inline snap/hooks/install:8]: Same here, quotes

File: snap/hooks/install [hunk 9]
```diff
@@ -2,58 +2,33 @@
...
+
+# Copy the bind config files to SNAP_COMMON
+cp -rf --preserve=mode "$SNAP/etc/bind" "$SNAP_COMMON/bind"
 
 # Fix the config files to take the new location into account
 SNAP_CURRENT="/var/snap/$SNAP_NAME/current"
-for file in $SNAP_DATA/etc/bind/* ; do
+for file in $SNAP_COMMON/bind/* ; do
```
Comment (arturo-seijas) [inline snap/hooks/install:15]: Same here, quotes

File: snap/hooks/install [hunk 10]
```diff
@@ -2,58 +2,33 @@
...
-  -e "s|<SNAP_COMMON>|${SNAP_COMMON}|g" \
-  "$SNAP_DATA/nginx/nginx.conf"
-
-# Create the static directory for django
-cp -r "$SNAP/api" "$SNAP_DATA/"
-chmod -R 755 $SNAP_DATA/api
-
-# Prepare the django app
+# Prepare the django database
 cd $SNAP_DATA/api
```
Comment (arturo-seijas) [inline snap/hooks/install:28]: I'd try to avoid cd-ing

File: snap/hooks/post-refresh [hunk 11]
```diff
@@ -0,0 +1,11 @@
+#!/bin/sh
+
+set -xe
+
+$SNAP/bin/prepare
+
+# Prepare the django database
+cd $SNAP_DATA/api
```
Comment (arturo-seijas) [inline snap/hooks/post-refresh:8]: Same here

File: service/bin/prepare (hunk 4 above)
Comment (nrobinaubertin) [reply service/bin/prepare:16]: This is coming in a subsequent PR: https://github.com/canonical/charmed-bind/pull/23

File: service/bin/prepare (hunk 5 above)
Comment (nrobinaubertin) [reply service/bin/prepare:30]: This is coming in a subsequent PR: https://github.com/canonical/charmed-bind/pull/23

File: service/bin/prepare (hunk 6 above)
Comment (nrobinaubertin) [reply service/bin/prepare:49]: This is coming in a subsequent PR: https://github.com/canonical/charmed-bind/pull/23

File: snap/hooks/install (hunk 9 above)
Comment (nrobinaubertin) [reply snap/hooks/install:15]: This is coming in a subsequent PR: https://github.com/canonical/charmed-bind/pull/23

File: api/dns/settings.py [hunk 12]
```diff
@@ -83,7 +83,7 @@
 DATABASES = {
     'default': {
         'ENGINE': 'django.db.backends.sqlite3',
-        'NAME': BASE_DIR / 'db.sqlite3',
+        'NAME': os.getenv("DJANGO_DATABASE_PATH", BASE_DIR / 'db.sqlite3'),
```
[the remark this answers is not in the document]
Comment (nrobinaubertin) [reply api/dns/settings.py]: I've changed that here but I need to add a linter for the rest of the django code

File: service/bin/prepare (hunk 7 above)
Comment (nrobinaubertin) [reply service/bin/prepare]: Fixed

File: snap/hooks/install (hunk 10 above)
Comment (nrobinaubertin) [reply snap/hooks/install]: Fixed

File: snap/hooks/post-refresh (hunk 11 above)
Comment (nrobinaubertin) [reply snap/hooks/post-refresh]: Fixed

File: snap/hooks/install (hunk 9 above)
Comment (arturo-seijas) [reply snap/hooks/install:15]: Ok. But it's simpler to review it when added/changed in the same PR so avoid some instances of this slipping unnoticed, though

Comment (arturo-seijas) [reply snap/hooks/install:15]: Makes sense

**Review by amandahla — APPROVED** [review]

File: service/bin/prepare [hunk 13]
```diff
@@ -0,0 +1,58 @@
+#!/bin/sh
+
+# The goal of this script is to prepare
```
Comment (javierdelapuente) [inline service/bin/prepare:3]: Not sure this sentence is finished

**Review by javierdelapuente — APPROVED** [review]
LGTM. As stated in other comments, waiting for shell linting in a following PR.

File: service/bin/prepare (hunk 13 above)
Comment (nrobinaubertin) [reply service/bin/prepare:3]: I'll remove that in the next PR

**Pull request merged**